### PR TITLE
Initial support for C++

### PR DESCRIPTION
**Pull request by frabert**
Missing in this PR:
  - `friend` declarations
  - Separate operations for methods, constructors and destructors (with support for the various attributes that those declarations have, like `virtual`, `volatile`, `const`, `&`, `&&`...)
  - Constructor initializers

### CODE REVIEW

File: include/vast/Translation/CodeGenContext.hpp [hunk 1]
```diff
@@ -130,7 +130,7 @@ namespace vast::cg
             return "anonymous[" + std::to_string(decl->getID()) + "]";
         }
 
-        std::string get_namespaced_for_decl_name(const clang::TagDecl *decl) {
+        std::string get_namespaced_for_decl_name(const clang::NamedDecl *decl) {
```
[the remark this answers is not in the document]
Comment (frabert) [reply include/vast/Translation/CodeGenContext.hpp:133]: `VarDecl`s also need to have a scoped name in the case of static variables inside records (I think).
`VarDecl`s are not `TagDecl`s, but both are `NamedDecl`s so this allows using the same logic for both cases

**Comment by lkorenc** [comment]
Also please move the newly added test to `HighLevel/CXX` (so it kind of mirrors the namespace `hl:cxx`).

File: include/vast/Dialect/HighLevel/HighLevelTypes.td [hunk 2]
```diff
@@ -272,6 +272,21 @@ class TypesOrLValueTypesMatch<string summary, string lhs, string rhs> : PredOpTr
   ]>
 >;
 
+//
+// RValue high-level type
+//
+def RValue : HighLevelType< "RValue",
```
Comment (lkorenc) [inline include/vast/Dialect/HighLevel/HighLevelTypes.td:278]: Copy & paste of `lvalue` extract that to shared helper.

File: include/vast/Dialect/HighLevel/HighLevelTypes.td [hunk 3]
```diff
@@ -442,4 +457,10 @@ def AdjustedType : HighLevelType< "Adjusted" > {
   let assemblyFormat = "`<` $original `,` $adjusted `>`";
 }
 
+def ReferenceType : HighLevelType< "Reference" > {
```
Comment (lkorenc) [inline include/vast/Dialect/HighLevel/HighLevelTypes.td:460]: Can use the same helper as lvalue/rvalue. Also should implement `SubElementTypeInterface`.

**Review by lkorenc — CHANGES_REQUESTED** [review]

**Review by lkorenc — APPROVED** [review]

**Pull request merged**